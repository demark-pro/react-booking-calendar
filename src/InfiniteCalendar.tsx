/** @jsxImportSource @emotion/react */
import React, {
  memo,
  useEffect,
  useMemo,
  useRef,
  createContext,
  useContext,
} from "react";

import format from "date-fns/format";
import isSameDay from "date-fns/isSameDay";
import isSameYear from "date-fns/isSameYear";

import {
  areEqual,
  FixedSizeGrid,
  GridChildComponentProps,
  GridOnScrollProps,
} from "react-window";
import AutoSizer from "react-virtualized-auto-sizer";

import {
  createDays,
  getDayState,
  getSelectedDates,
  getStyleProps,
} from "./helpers";
import {
  CalendarContextProps,
  GridProps,
  Day,
  CommonPropsType,
  DefaultCalendarProps,
  DayState,
} from "./types";

import { checkOberbooking } from "./utils/checkOberbooking";
import {
  dateFnsOptionsInit,
  dateOfStartMonthInit,
  disabledInit,
  isStartInit,
  rangeInit,
  reservedInit,
  selectedInit,
  variantInit,
} from "./Calendar";
import { defaultComponents } from "./components";
import { DayCellProps } from "./components/DayCell";
import { defaultStyles, StylesProps } from "./styles";

export type InfiniteCalendarProps = DefaultCalendarProps & {
  numOfMonths?: number;
  colHeight?: number;
  scrollToDate?: Date | number | null;
  onScroll?: (e: GridOnScrollProps) => void;
};

// ==============================
// Context
// ==============================

const CalendarContext = createContext<CalendarContextProps | null>(null);
const useProps = () => {
  const currentCalendarContext = useContext(CalendarContext);

  if (!currentCalendarContext) {
    throw new Error(
      "useComponent has to be used within <CalendarContext.Provider>"
    );
  }

  return currentCalendarContext;
};

// ==============================
// RenderCell
// ==============================

const RenderCell = memo((props: GridChildComponentProps): JSX.Element => {
  const { columnIndex, rowIndex, data, style } = props;
  const index = 7 * rowIndex + columnIndex;
  let day = data[index];

  const { components, commonProps, handleClickDay } = useProps();

  const { selected, reserved, dateFnsOptions } = commonProps;

  const { DayCell, DayCellContent, DayCellFooter, DayCellHeader } = components;

  const state = getDayState({ day, selected, reserved });
  const dayProps: DayCellProps & CommonPropsType = {
    date: day.date,
    state,
    ...commonProps,
  };

  if (!DayCell || !DayCellContent || !DayCellFooter || !DayCellHeader)
    return <></>;

  // empty col
  if (!state.isSameMonth) {
    dayProps.state = { isSameMonth: false };
    return <DayCell innerProps={{ style: style }} {...dayProps} />;
  }

  return (
    <DayCell
      innerProps={{
        style: style,
        onClick: () => handleClickDay(day, state),
      }}
      {...dayProps}
    >
      <DayCellHeader {...dayProps} />
      <DayCellContent {...dayProps} />
      <DayCellFooter {...dayProps} />
      {state.isStartMonth && (
        <div
          {...getStyleProps({}, "month", commonProps)}
          style={{ top: state.isSameYear ? "-30px" : "-40px" }}
        >
          <span style={{ fontWeight: 600, color: "#000!important" }}>
            {format(day.date, "LLL", dateFnsOptions)}
          </span>
          <small>
            {!isSameYear(day.date, new Date()) && format(day.date, "yyyy")}
          </small>
        </div>
      )}
    </DayCell>
  );
}, areEqual);

// ==============================
// Grid
// ==============================

const Grid = ({
  width,
  height,
  items,
  colHeight,
  scrollToDate,
  onScroll,
  ...props
}: GridProps) => {
  const listRef = useRef<FixedSizeGrid>(null);

  useEffect(() => {
    if (!scrollToDate || !items?.length) return;

    const day = items.find(
      (d) => isSameDay(d.date, scrollToDate) && d.rowIndex > -1
    );

    if (!!day && listRef?.current) {
      listRef.current.scrollToItem({
        align: "start",
        rowIndex: day.rowIndex,
      });
    }
  }, [scrollToDate, items]);

  return (
    <FixedSizeGrid
      ref={listRef}
      itemData={items}
      columnCount={7}
      columnWidth={width / 7}
      height={height}
      rowCount={items.length / 7}
      rowHeight={colHeight}
      width={width}
      overscanRowCount={3}
      onScroll={onScroll}
      {...props}
    >
      {RenderCell}
    </FixedSizeGrid>
  );
};

// ==============================
// InfiniteCalendar
// ==============================

function InfiniteCalendar(props: InfiniteCalendarProps) {
  const {
    selected = selectedInit,
    isStart = isStartInit,
    dateOfStartMonth = dateOfStartMonthInit,
    reserved = reservedInit,
    variant = variantInit,
    dateFnsOptions = dateFnsOptionsInit,
    range = rangeInit,
    disabled = disabledInit,
    components = {},
    scrollToDate = null,
    numOfMonths = 12,
    colHeight = 55,
    className = "",
    classNamePrefix,
    onOverbook,
    onChange,
    onScroll,
    ...innerProps
  } = props;
  const weekHeight = colHeight / 2;

  const [startDate, endDate] = getSelectedDates(selected);

  const handleClickDay = (day: Day, state: DayState) => {
    if (typeof disabled === "boolean" && disabled) return;

    const { date } = day;
    const { errorType, newSelected } = checkOberbooking({
      date,
      selected,
      isStart,
      reserved,
      state,
      range,
      variant,
    });

    if (errorType && !newSelected) {
      if (onOverbook) onOverbook(date, errorType);
      return;
    }

    if (newSelected && onChange) onChange(newSelected);
  };

  const startMonth = format(dateOfStartMonth, "yyyy-MM");
  const items = useMemo(
    () =>
      createDays({
        dateOfStartMonth,
        numOfMonths,
        dateFnsOptions,
      }),
    [startMonth, numOfMonths, dateFnsOptions]
  );

  // getClassNames

  const getClassNames = (name: string) => {
    if (!classNamePrefix) return "";

    const prefix = classNamePrefix.length ? classNamePrefix + "__" : "";

    return prefix + name.replaceAll("_", "-");
  };

  // getStyles

  const getStyles = <Key extends keyof StylesProps>(props: any, name: Key) => {
    const base = defaultStyles[name](props as any);
    base.boxSizing = "border-box";

    return base;
  };

  // commonProps

  const commonProps = {
    selected: [startDate, endDate],
    reserved,
    variant,
    dateFnsOptions,
    getClassNames,
    getStyles,
  };

  const currentComponents = defaultComponents(components);
  const { CalendarContainer, WeekContainer, WeekCell } = currentComponents;

  return (
    <CalendarContext.Provider
      value={{
        components: currentComponents,
        commonProps,
        handleClickDay,
      }}
    >
      <CalendarContainer {...commonProps} innerProps={innerProps}>
        <WeekContainer
          innerProps={{ style: { height: weekHeight } }}
          {...commonProps}
        >
          {WeekCell}
        </WeekContainer>
        <AutoSizer>
          {({ height, width }) => (
            <Grid
              height={height - weekHeight}
              width={width}
              items={items}
              colHeight={colHeight}
              scrollToDate={scrollToDate}
              onScroll={onScroll}
            />
          )}
        </AutoSizer>
      </CalendarContainer>
    </CalendarContext.Provider>
  );
}

export default InfiniteCalendar;
