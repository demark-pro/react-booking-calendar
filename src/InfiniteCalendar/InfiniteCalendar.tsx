import React, {
  memo,
  useState,
  useEffect,
  useMemo,
  useRef,
  createContext,
  useContext,
} from "react";

import isBefore from "date-fns/isBefore";
import format from "date-fns/format";
import differenceInCalendarWeeks from "date-fns/differenceInCalendarWeeks";
import differenceInCalendarMonths from "date-fns/differenceInCalendarMonths";
import startOfDay from "date-fns/startOfDay";
import addWeeks from "date-fns/addWeeks";
import startOfMonth from "date-fns/startOfMonth";
import startOfWeek from "date-fns/startOfWeek";
import addDays from "date-fns/addDays";
import formatRelative from "date-fns/formatRelative";
import isSameDay from "date-fns/isSameDay";
import isSameYear from "date-fns/isSameYear";

import { areEqual, FixedSizeGrid, GridChildComponentProps } from "react-window";
import AutoSizer from "react-virtualized-auto-sizer";

import { createDays, isBetween, isBetweenInterval } from "../helpers";
import { getSelectedTime } from "../utils/getSelectedTime";
import {
  CalendarContextProps,
  InfiniteCalendarProps,
  GridProps,
  DateFnsOptions,
  DayInfo,
  Reserved,
  Titles,
  DayInfoFooter,
  DayInfoHeader,
  DayInfoMonth,
} from "../types";

import styles from "./InfiniteCalendar.module.css";

const PAST = "PAST";
const BOOKED = "BOOKED";
const BEFORE_START = "BEFORE_START";
const AFTER_END = "AFTER_END";
const BOOKED_BETWEEN = "BOOKED_BETWEEN";

const titlesInit: Titles = {
  dayFooterStart: "check-in",
  dayFooterEnd: "exit",
  reservedFooter: "booked",
};

const dateFnsOptionsInit: DateFnsOptions = {
  weekStartsOn: 1,
};

const CalendarContext = createContext<CalendarContextProps>({
  selectedDates: [null, null],
  dateFnsOptions: dateFnsOptionsInit,
  titles: titlesInit,
  handleClickDay: () => {},
});

function checkOberbooking(
  day: DayInfo,
  selectedDates: Array<Date | number | null>,
  isStart: boolean,
  reserved: Reserved[]
): string | null {
  const [selectedStart, selectedEnd] = selectedDates;
  const { date, isReserved, isPast } = day;

  if (isPast) return PAST;

  if (isReserved) return BOOKED;

  if (!isStart && selectedStart && isBefore(date, selectedStart))
    return BEFORE_START;

  if (isStart && !selectedStart && selectedEnd && isBefore(selectedEnd, date))
    return AFTER_END;

  const isReservedBetween =
    selectedStart &&
    !!reserved.find((r) =>
      isBetweenInterval(r.startDate, r.endDate, selectedStart, startOfDay(date))
    );
  if (isReservedBetween) return BOOKED_BETWEEN;

  return null;
}

const Cell = memo((props: GridChildComponentProps): JSX.Element => {
  const { columnIndex, rowIndex, data, style } = props;
  const index = 7 * rowIndex + columnIndex;
  let day = { ...data[index], classNames: [styles.day_col] };

  const {
    selectedDates,
    dateFnsOptions,
    titles,
    renderDay,
    handleClickDay,
  } = useContext(CalendarContext);
  const [selectedStart, selectedEnd] = selectedDates;

  // prettier-ignore
  const isSelectedStart = selectedStart ? isSameDay(day.date, selectedStart) : false;
  // prettier-ignore
  const isSelectedEnd = selectedEnd ? isSameDay(day.date, selectedEnd) : false;
  // prettier-ignore
  const isSelected = !!(selectedStart && selectedEnd && isBetween(day.date, selectedStart, selectedEnd, "{}"));
  if (isSelected) day.classNames.push(styles.selected);
  if (isSelectedStart) day.classNames.push(styles.selected_start);
  if (isSelectedEnd) day.classNames.push(styles.selected_end);
  if (day.isReserved) day.classNames.push(styles.reserved);
  if (day.isToday) day.classNames.push(styles.today);
  if (day.isPast) day.classNames.push(styles.past);

  // header
  let header: DayInfoHeader = {
    classNames: [styles.day_col_header],
    text: "",
    visible: false,
  };
  // footer
  let footer: DayInfoFooter = {
    classNames: [styles.day_col_footer],
    text: "",
    visible: false,
  };
  // month
  let month: DayInfoMonth | null = null;

  if (day.isStartMonth) {
    month = {
      classNames: [styles.month],
      text: format(day.date, "LLL", dateFnsOptions),
      yearText: !isSameYear(day.date, new Date())
        ? format(day.date, "yyyy")
        : null,
    };
  }

  if (isSelectedStart || isSelectedEnd) {
    header.visible = true;
    header.text = format(day.date, "MMM", dateFnsOptions);
    footer.text = isSelectedStart ? titles.dayFooterStart : titles.dayFooterEnd;
  }

  if (day.isToday) {
    header.visible = true;
    header.text = formatRelative(new Date(), new Date(), dateFnsOptions).split(
      " "
    )[0];
  }

  // prettier-ignore
  const reservedDate = day.reservedStart || day.reservedEnd

  if (day.isReserved) footer.text = titles.reservedFooter;
  if (reservedDate) footer.text = format(reservedDate, "HH:mm");

  day = { ...day, header, footer, month };

  // empty col
  if (!day.isSameMonth) {
    return (
      <div
        key={format(day.date, "dd-MM-yyyy-empty")}
        className={styles.empty_day}
        style={style}
      />
    );
  }

  if (renderDay) return renderDay({ ...day, style });

  return (
    <div
      key={format(day.date, "dd-MM-yyyy")}
      className={day.classNames.join(" ")}
      style={style}
      onClick={() => handleClickDay(day)}
    >
      {day.month && (
        <div
          className={day.month.classNames.join(" ")}
          style={{ top: day.isSameYear ? "-30px" : "-50px" }}
        >
          <span style={{ fontWeight: 600, color: "#000!important" }}>
            {day.month.text}
          </span>
          <small>{day.month.yearText && day.month.yearText}</small>
        </div>
      )}
      {day.header.visible && (
        <span className={day.header.classNames.join(" ")}>
          {day.header.text}
        </span>
      )}
      <span className={styles.day_col_date}>{day.text}</span>
      <span className={day.footer.classNames.join(" ")}>{day.footer.text}</span>
      {(isSelectedStart || isSelectedEnd) && (
        <div className={styles.selected_event} />
      )}
      {isSelected && (
        <div
          className={styles.selected_between}
          style={{
            left: isSelectedStart ? "50%" : 0,
            right: isSelectedEnd ? "50%" : 0,
          }}
        />
      )}
    </div>
  );
}, areEqual);

function Grid({
  dateOfStartMonth,
  overscanWeekCount,
  width,
  height,
  items,
  colHeight,
  scrollToDate,
  onScroll,
}: GridProps) {
  const listRef = useRef<FixedSizeGrid>(null);

  useEffect(() => {
    if (!scrollToDate || !items?.length) return;

    const day = items.find(
      (d) => isSameDay(d.date, scrollToDate) && d.rowIndex > -1
    );

    if (!!day && listRef?.current) {
      console.log(day)
      listRef.current.scrollToItem({
        align: "start",
        rowIndex: day.rowIndex,
      });
    }

    const diffWeeks = differenceInCalendarWeeks(
      scrollToDate,
      startOfMonth(dateOfStartMonth)
    );
    const diffMonths =
      differenceInCalendarMonths(scrollToDate, startOfMonth(dateOfStartMonth)) *
      2;
  }, [scrollToDate]);

  const calcWeekWithMonths = useMemo(
    () =>
      overscanWeekCount +
      differenceInCalendarMonths(
        dateOfStartMonth,
        addWeeks(dateOfStartMonth, overscanWeekCount)
      ) +
      2,
    [overscanWeekCount]
  );

  return (
    <FixedSizeGrid
      ref={listRef}
      itemData={items}
      columnCount={7}
      columnWidth={width / 7}
      height={height}
      rowCount={items.length / 7}
      rowHeight={colHeight}
      width={width}
      className={styles.grid}
      overscanRowCount={calcWeekWithMonths}
      onScroll={onScroll}
    >
      {Cell}
    </FixedSizeGrid>
  );
}

function InfiniteCalendar({
  selectedStart = null,
  selectedEnd = null,
  isStart = true,
  dateOfStartMonth = new Date(),
  numOfMonths = 12,
  overscanWeekCount = 4,
  colHeight = 55,
  reserved = [],
  titles = titlesInit,
  disabled = false,
  scrollToDate = null,
  dateFnsOptions = dateFnsOptionsInit,
  rangeMode = false,
  renderDay,
  onOverbook,
  onChange,
  onChangeRange,
  onScroll,
  className = "",
  ...props
}: InfiniteCalendarProps) {
  const weekHeight = colHeight / 2;
  const startMonth = format(dateOfStartMonth, "yyyy-MM");

  const [selectedDates, setSelectedDates] = useState<
    Array<Date | number | null>
  >([null, null]);
  const [startDate, endDate] = selectedDates;

  useEffect(() => {
    setSelectedDates([selectedStart, selectedEnd]);
  }, [selectedStart, selectedEnd]);

  const renderWeeks = () => {
    const dateFormat = "EEEEEE";
    const weeks = [];
    let startDate = startOfWeek(new Date(), dateFnsOptions);
    for (let i = 0; i < 7; i++) {
      weeks.push(
        <div
          key={i}
          className={styles.week_col}
          style={{ color: i > 4 ? "red" : "#424242" }}
        >
          {format(addDays(startDate, i), dateFormat, dateFnsOptions)}
        </div>
      );
    }

    return (
      <div className={styles.week_row} style={{ height: weekHeight }}>
        {weeks}
      </div>
    );
  };

  const handleClickDay = (day: DayInfo) => {
    if (disabled) return;
    const { date } = day;

    const overbookError = checkOberbooking(
      day,
      selectedDates,
      rangeMode ? !startDate : isStart,
      reserved
    );

    if (rangeMode) {
      if ((!startDate && !endDate) || (startDate && endDate)) {
        if (overbookError !== BOOKED) {
          setSelectedDates([date, null]);
          return;
        }
      }
    }

    if (overbookError) {
      if (rangeMode) {
        if (
          overbookError === BOOKED_BETWEEN ||
          overbookError === BEFORE_START
        ) {
          setSelectedDates([date, null]);
          return;
        }
      }

      if (onOverbook) onOverbook(date, overbookError);
      return;
    }

    const dayWithTime = getSelectedTime(date, reserved, startDate);

    if (rangeMode) {
      if (onChangeRange) {
        onChangeRange([startDate || dayWithTime, endDate || dayWithTime]);
      }
      return;
    }

    if (onChange && dayWithTime) onChange(dayWithTime);
  };

  const items = useMemo(
    () =>
      createDays({
        dateOfStartMonth,
        numOfMonths,
        reserved,
      }),
    [startMonth, numOfMonths, reserved, titles, dateFnsOptions]
  );

  return (
    <CalendarContext.Provider
      value={{
        selectedDates,
        dateFnsOptions,
        titles,
        renderDay,
        handleClickDay,
      }}
    >
      <div className={styles.calendar + " " + className} {...props}>
        {renderWeeks()}
        <AutoSizer>
          {({ height, width }) => (
            <Grid
              dateOfStartMonth={dateOfStartMonth}
              overscanWeekCount={overscanWeekCount}
              height={height - weekHeight}
              width={width}
              items={items}
              colHeight={colHeight}
              scrollToDate={scrollToDate}
              onScroll={onScroll}
            />
          )}
        </AutoSizer>
      </div>
    </CalendarContext.Provider>
  );
}

export default InfiniteCalendar;
