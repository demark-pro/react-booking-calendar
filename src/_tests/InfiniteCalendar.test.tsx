import React from "react";
import { render } from "@testing-library/react";

import InfiniteCalendar from "../InfiniteCalendar";

jest.mock("./InfiniteCalendar.module.css", () => "");
describe("InfiniteCalendar", () => {
  test("renders the InfiniteCalendar component", () => {
    render(<InfiniteCalendar selected={[new Date(), null]} />);
  });
});
