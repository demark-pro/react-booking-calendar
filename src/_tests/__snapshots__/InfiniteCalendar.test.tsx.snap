// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot - InfiniteCalendar 1`] = `
.emotion-0 {
  background-color: #fff;
  position: relative;
  height: 100%;
  width: 100%;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen","Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue",sans-serif;
  -webkit-font-smoothing: antialiased;
  box-sizing: border-box;
}

.emotion-1 {
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  left: 0;
  right: 0;
  background: #fff;
  z-index: 10;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  width: 100%;
  box-sizing: border-box;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-basis: calc(100% / 7);
  -ms-flex-preferred-size: calc(100% / 7);
  flex-basis: calc(100% / 7);
  width: calc(100% / 7);
  font-size: 10px;
  text-transform: uppercase;
  padding: 0.5rem 0;
  font-weight: 700;
  color: red;
  box-sizing: border-box;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-basis: calc(100% / 7);
  -ms-flex-preferred-size: calc(100% / 7);
  flex-basis: calc(100% / 7);
  width: calc(100% / 7);
  font-size: 10px;
  text-transform: uppercase;
  padding: 0.5rem 0;
  font-weight: 700;
  color: #424242;
  box-sizing: border-box;
}

<div>
  <div
    class="infinite-calendar__calendar-container emotion-0"
  >
    <div
      class="infinite-calendar__week-container emotion-1"
      style="height: 27.5px;"
    >
      <div
        class="infinite-calendar__weekCell emotion-2"
      >
        Su
      </div>
      <div
        class="infinite-calendar__weekCell emotion-3"
      >
        Mo
      </div>
      <div
        class="infinite-calendar__weekCell emotion-3"
      >
        Tu
      </div>
      <div
        class="infinite-calendar__weekCell emotion-3"
      >
        We
      </div>
      <div
        class="infinite-calendar__weekCell emotion-3"
      >
        Th
      </div>
      <div
        class="infinite-calendar__weekCell emotion-3"
      >
        Fr
      </div>
      <div
        class="infinite-calendar__weekCell emotion-2"
      >
        Sa
      </div>
    </div>
    <div
      style="overflow: visible; height: 0px; width: 0px;"
    />
    <div
      class="resize-triggers"
    >
      <div
        class="expand-trigger"
      >
        <div
          style="width: 1px; height: 1px;"
        />
      </div>
      <div
        class="contract-trigger"
      />
    </div>
  </div>
</div>
`;
